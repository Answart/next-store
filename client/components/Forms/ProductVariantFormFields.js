import PropTypes from 'prop-types';
import NProgress from 'nprogress';
import StyledProduct from '../styles/ProductStyles';
import { COLORS, SIZES } from '../../config';
import { uploadImageFile, destroyImageFileByToken } from '../../lib/cloudinary';


const ProductVariantFormFields = props => {
  const { sale, image, imageIsNew } = props;
  const handleChange = e => {
    if (!!e.preventDefault) e.preventDefault();
    const { name, type, value, checked } = e.target;
    let val = value;
    let state = {};

    if (type === 'number') {
      val = (name == 'price' || name === 'salePrice')
        ? Number(parseFloat(value).toFixed(2))
        : Number(parseFloat(value));
    }
    if (type === 'checkbox') val = checked;
    if (name === 'sale') state.salePrice = 1.00;

    state[name] = val;
    props.saveToForm(state);
  };
  const handleImageChange = async e => {
    if (!!e.preventDefault) e.preventDefault();
    const { name, type, value, files } = e.target;
    const currentImageToken = props.image
      ? (props.image.delete_token || '')
      : '';
    NProgress.start();

    if (type === 'radio') {
      const imageIsNew = (value === 'true');
      if (imageIsNew && !!currentImageToken.length) await destroyImageFileByToken(currentImageToken);

      props.saveToForm({ imageIsNew });
    }
    if (type === 'file' && name === 'upload') {
      if (!files || !files.length) return;
      const image = await uploadImageFile(files[0]);
      if (image.error) return alert('An error occured while uploading image. Please try again later.');
      if (!!currentImageToken.length) await destroyImageFileByToken(currentImageToken);

      props.saveToForm({ image });
    }
    NProgress.done();
  }
  return (
    <StyledProduct>
      <div className="form-imgs">
        {!!image ? (
          <img width="450" height="640" src={image.large_image_url} alt={image.name} />
        ) : (
          <img width="450" height="640" src="/static/images/placeholder_large.jpg" alt="Placeholder Image" />
        )}
      </div>

      <div className="form-content">
        <div className="field-padding">
          <label htmlFor="size">
            Size:
            <select id="size"
              name="size"
              disabled={props.editView}
              value={props.size}
              onChange={handleChange}
            >
              <option key={0} value=''></option>
              {SIZES.map(sz => <option key={sz} value={sz}>{sz}</option>)}
            </select>
          </label>
        </div>

        <div className="field-padding">
          <label htmlFor="color">
            Color:
            <select id="color"
              name="color"
              disabled={props.editView}
              value={props.color}
              onChange={handleChange}
            >
              <option key={0} value=''></option>
              {COLORS.map(clr => <option key={clr} value={clr}>{clr}</option>)}
            </select>
          </label>
        </div>

        <div className="field-padding">
          <label htmlFor="quantity">
            Quantity:
            <input id="quantity"
              type="number"
              name="quantity"
              placeholder="1"
              min="1"
              value={props.quantity}
              onChange={handleChange}
              required
            />
          </label>
        </div>

        <div className="field-padding">
          <label htmlFor="price">
            Price per Item (USD):
            <input id="price"
              type="number"
              name="price"
              placeholder={1.00}
              min={1.00}
              step={0.01}
              value={props.price}
              onChange={handleChange}
              required
            />
          </label>
        </div>

        <div className="field-padding">
          <label htmlFor="sale">
            <input id="sale"
              type="checkbox"
              name="sale"
              key={sale}
              value={sale}
              onChange={handleChange}
              checked={sale ? "checked" : ""}
            />On Sale
          </label>
        </div>

        {sale && (
          <div className="field-padding">
            <label htmlFor="salePrice">
              Sale Price per Item (USD):
              <input id="salePrice"
                type="number"
                name="salePrice"
                placeholder={1.00}
                min={1.00}
                step={0.01}
                disabled={!sale}
                value={props.salePrice}
                onChange={handleChange}
              />
            </label>
          </div>
        )}

        <div className="field-padding">
          <label htmlFor="image">
            Image:
            <div className="prdct-padding">
              <div className="prdct-padding">
                <label htmlFor="productImage" className="inline-lbl">
                  <input id="productImage"
                    type="radio"
                    name="productImage"
                    key={!imageIsNew}
                    value={false}
                    onChange={handleImageChange}
                    checked={!imageIsNew}
                  />Same as Product Image
                </label>
              </div>
              <div>
                <label htmlFor="newImage" className="inline-lbl">
                  <input id="newImage"
                    type="radio"
                    name="newImage"
                    key={imageIsNew}
                    value={true}
                    onChange={handleImageChange}
                    checked={imageIsNew}
                  />New Image
                </label>
              </div>

              {imageIsNew && (
                <div className="field-padding">
                  <label htmlFor="upload" className="lbl-button">
                    <input id="upload"
                      type="file"
                      name="upload"
                      onChange={handleImageChange}
                    />Upload Image
                  </label>
                </div>
              )}

              {!!image && !!image.name.length && (
                <div className="field-padding field-detail">
                  <h5>
                    Upload Details
                  </h5>
                  <p>
                    ID: <i>{image.cloudinary_id}</i>
                  </p>
                  <p>
                    Name: <i>{image.name}</i>
                  </p>
                  <p>
                    Dimensions: <i>{image.width}</i>W x <i>{image.height}</i>H
                  </p>
                  <p>
                    Transformation: <i>{image.transformation}</i>
                  </p>
                  <p>
                    Url: <i>{image.image_url}</i>
                  </p>
                </div>
              )}
            </div>
          </label>
        </div>
      </div>
    </StyledProduct>
  );
};

ProductVariantFormFields.propTypes = {
  price: PropTypes.number.isRequired,
  quantity: PropTypes.number.isRequired,
  size: PropTypes.string,
  color: PropTypes.string,
  sale: PropTypes.bool.isRequired,
  salePrice: PropTypes.number.isRequired,
  image: PropTypes.shape({
    id: PropTypes.string,
    cloudinary_id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
    transformation: PropTypes.string.isRequired,
    image_url: PropTypes.string.isRequired,
    large_image_url: PropTypes.string.isRequired
  }),
  imageIsNew: PropTypes.bool.isRequired,
  saveToForm: PropTypes.func.isRequired,
  editView: PropTypes.bool.isRequired
};


export default ProductVariantFormFields;
