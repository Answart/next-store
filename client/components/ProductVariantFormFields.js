import React, { Component } from 'react';
import PropTypes from 'prop-types';
import colors from '../lib/colors';
import sizes from '../lib/sizes';


class ProductVariantFormFields extends Component {
  static propTypes = {
    price: PropTypes.number.isRequired,
    quantity: PropTypes.number.isRequired,
    size: PropTypes.string,
    color: PropTypes.string,
    sale: PropTypes.bool.isRequired,
    salePrice: PropTypes.number,
    handleChange: PropTypes.func.isRequired
  };
  render() {
    const {
      price,
      quantity,
      size,
      color,
      sale,
      salePrice,
      handleChange
    } = this.props;
    return (
      <div>
        <label htmlFor="size">
          Size:
          <select
            id="size"
            name="size"
            value={size}
            onChange={handleChange}
          >
            <option key={0} value={''}></option>
            {sizes.map(sz => <option key={sz} value={sz}>{sz}</option>)}
          </select>
        </label>

        <label htmlFor="color">
          Color:
          <select
            id="color"
            name="color"
            placeholder=""
            value={color}
            onChange={handleChange}
          >
            <option key={0} value={''}></option>
            {colors.map(clr => <option key={clr} value={clr}>{clr}</option>)}
          </select>
        </label>

        <label htmlFor="quantity">
          Quantity:
          <input
            type="number"
            id="quantity"
            name="quantity"
            placeholder="Quantity"
            min="1"
            value={quantity}
            onChange={handleChange}
            required
          />
        </label>

        <label htmlFor="price" className="prdct-padding">
          Price per Item:
          <input
            type="number"
            id="price"
            name="price"
            placeholder="42.99"
            min="1"
            value={price}
            onChange={handleChange}
            required
          />
        </label>

        <div>
          <input
            type="checkbox"
            id="sale"
            name="sale"
            value={sale}
            onChange={handleChange}
            checked={sale ? "checked" : ""}
          />
          <label htmlFor="sale" className="prdct-padding">
            Is this selection on sale?
          </label>
        </div>

        {sale && (
          <label htmlFor="salePrice" className="prdct-padding">
            Sale Price per Item:
            <input
              type="number"
              id="salePrice"
              name="salePrice"
              placeholder="22.50"
              min="1"
              disabled={!sale}
              value={salePrice}
              onChange={handleChange}
            />
          </label>
        )}
      </div>
    );
  }
}

export default ProductVariantFormFields;
