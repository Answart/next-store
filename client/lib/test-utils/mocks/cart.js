import {
  ADD_TO_CART_MUTATION,
  UPDATE_CARTITEM_MUTATION,
  REMOVE_FROM_CART_MUTATION
} from '../../../graphql'
import { fakeCartItem } from '../utils';
import { mockVariant } from './variant';

const mockCartItem = fakeCartItem();


const addToCartMutationMock = {
  request: { query: ADD_TO_CART_MUTATION, variables: { id: mockVariant.id } },
  result: {
    data: {
      addToCart: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
      },
    },
  },
};

const updateCartItemMutationMock = (quantity) => ({
  request: { query: UPDATE_CARTITEM_MUTATION, variables: { id: mockCartItem.id, quantity } },
  result: {
    data: {
      updateCartItem: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
        quantity,
        variant: { ...mockCartItem.variant }
      },
    },
  },
});

const removeFromCartMutationMock = {
  request: { query: REMOVE_FROM_CART_MUTATION, variables: { id: mockCartItem.id } },
  result: {
    data: {
      removeFromCart: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
      },
    },
  },
};


export {
  mockCartItem,
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock
}
