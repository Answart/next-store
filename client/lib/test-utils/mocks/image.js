import { CREATE_IMAGE_MUTATION } from '../../../graphql'
import { fakeImage } from '../utils';

const mockImage = fakeImage();
const mockImageVariables = {
  cloudinary_id: mockImage.cloudinary_id,
  name: mockImage.name,
  height: mockImage.height,
  width: mockImage.width,
  transformation: mockImage.transformation,
  image_url: mockImage.image_url,
  large_image_url: mockImage.large_image_url
}


const createImageMutationMock = (overrides) => ({
  request: {
    query: CREATE_IMAGE_MUTATION,
    variables: {
      ...mockImageVariables,
      ...overrides,
    },
  },
  result: {
    data: {
      createImage: {
        __typename: 'Image',
        id: mockImage.id,
      },
    },
  },
});

const createImageMutationErrorMock = {
  request: {
    query: CREATE_IMAGE_MUTATION,
    variables: {
      ...mockImageVariables,
      cloudinary_id: '',
      name: '',
    },
  },
  result: {
    errors: [{ message: 'ack!' }],
  }
};


export {
  mockImage, mockImageVariables,
  createImageMutationMock, createImageMutationErrorMock,
}
