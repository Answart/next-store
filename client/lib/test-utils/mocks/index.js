export {
  mockUser, mockUsers,
  mockImage, mockImageVariables,
  mockProduct, mockProducts, mockShopProductsVariables,
  mockVariant,
  mockCartItem,
  mockOrderItem, mockOrder,
} from './typeDefs';
export {
  localUserQueryEmptyCartMock, localUserQueryNoUserMock, localUserQueryCartItemMock,
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signupMutationMock,
  signinMutationMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock,
  resetPasswordMutationMock, resetPasswordMutationErrorMock,
  updatePermissionsMutationMock,
  createImageMutationMock, createImageMutationErrorMock,
  productQueryMock, productQueryErrorMock, productQueryNoProductMock, productQueryNoVariantMock,
  shopProductsQueryFilterMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
  createProductVariantMutationMock, createProductVariantMutationErrorMock,
  updateProductVariantMutationMock, updateProductVariantMutationErrorMock,
  deleteProductVariantMutationMock,
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock,
} from './resolvers';
