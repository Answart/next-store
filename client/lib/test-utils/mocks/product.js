import {
  PRODUCT_QUERY, SHOP_PRODUCTS_QUERY,
  CREATE_PRODUCT_MUTATION,
  UPDATE_PRODUCT_MUTATION,
  DELETE_PRODUCT_MUTATION,
} from '../../../graphql'
import { mockUser, mockProduct, mockImage, mockVariant, mockShopProductsVariables } from '../utils';


const productQueryMock = {
  request: { query: PRODUCT_QUERY, variables: { id: mockProduct.id } },
  result: {
    data: {
      product: {
        ...mockProduct,
        __typename: mockProduct.__typename,
        image: mockImage,
        user: {
          __typename: mockUser.__typename,
          id: mockUser.id,
          name: mockUser.name,
        },
        variants: [{
          ...mockVariant,
          product: {
            __typename: mockProduct.__typename,
            id: mockProduct.id,
            image: mockImage
          }
        }]
      },
    },
  },
};

const productQueryNoVariantMock = {
  request: { query: PRODUCT_QUERY, variables: { id: mockProduct.id } },
  result: {
    data: {
      product: {
        ...mockProduct,
        __typename: mockProduct.__typename,
        image: mockImage,
        user: {
          __typename: mockUser.__typename,
          id: mockUser.id,
          name: mockUser.name,
        },
        variants: []
      },
    },
  },
};

const createProductMutationMock = {
  request: {
    query: CREATE_PRODUCT_MUTATION,
    variables: {
      department: mockProduct.department,
      title: mockProduct.title,
      description: mockProduct.description,
      category: mockProduct.category,
      brand: mockProduct.brand,
      online: mockProduct.online,
      imageId: mockImage.id,
    },
  },
  result: {
    data: {
      createProduct: {
        __typename: 'Product',
        id: mockProduct.id,
      },
    },
  },
};

const createProductMutationErrorMock = {
  request: {
    query: CREATE_PRODUCT_MUTATION,
    variables: {
      department: 'tops',
      title: '',
      description: '',
      category: '',
      brand: '',
      online: false,
      imageId: mockImage.id,
    },
  },
  result: {
    errors: [{ message: 'ack!' }],
  }
};

const updateProductMutationMock = {
  request: {
    query: UPDATE_PRODUCT_MUTATION,
    variables: {
      id: mockProduct.id,
      department: mockProduct.department,
      title: mockProduct.title,
      description: mockProduct.description,
      category: mockProduct.category,
      brand: mockProduct.brand,
      online: mockProduct.online,
      imageId: mockImage.id,
    },
  },
  result: {
    data: {
      updateProduct: {
        __typename: 'Product',
        id: mockProduct.id,
      },
    },
  },
};

const updateProductMutationErrorMock = {
  request: {
    query: UPDATE_PRODUCT_MUTATION,
    variables: {
      id: mockProduct.id,
      department: 'tops',
      title: '',
      description: '',
      category: '',
      brand: '',
      online: false,
      imageId: mockImage.id,
    },
  },
  result: {
    errors: [{ message: 'ack!' }],
  }
};

const deleteProductMutationMock = {
  request: { query: DELETE_PRODUCT_MUTATION, variables: { id: mockProduct.id } },
  result: {
    data: {
      deleteProduct: {
        __typename: 'Product',
        id: mockProduct.id,
      },
    },
  },
};

const shopProductsQueryNameEmptyMock = {
  request: { query: SHOP_PRODUCTS_QUERY, variables: { name: mockUser.name }},
  result: {
    data: {
      products: [],
    },
  },
};

const shopProductsQueryProductMock = {
  request: { query: SHOP_PRODUCTS_QUERY, variables: { ...mockShopProductsVariables }},
  result: {
    data: {
      products: [{
        ...mockProduct,
        image: mockImage,
        user: {
          __typename: mockUser.__typename,
          id: mockUser.id,
          name: mockUser.name,
        },
        variants: [{
          ...mockVariant,
          product: {
            __typename: mockProduct.__typename,
            id: mockProduct.id,
            image: mockImage,
          }
        }]
      }],
    },
  },
};


export {
  productQueryMock, productQueryNoVariantMock,
  shopProductsQueryNameEmptyMock, shopProductsQueryProductMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
}
