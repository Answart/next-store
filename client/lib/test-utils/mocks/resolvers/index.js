export {
  localUserQueryEmptyCartMock, localUserQueryNoUserMock, localUserQueryCartItemMock,
} from './local';
export {
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signupMutationMock,
  signinMutationMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock,
  resetPasswordMutationMock, resetPasswordMutationErrorMock,
  updatePermissionsMutationMock,
} from './user';
export {
  createImageMutationMock, createImageMutationErrorMock,
} from './image';
export {
  productQueryMock, productQueryErrorMock, productQueryNoProductMock, productQueryNoVariantMock,
  shopProductsQueryFilterMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
} from './product';
export {
  createProductVariantMutationMock, createProductVariantMutationErrorMock,
  updateProductVariantMutationMock, updateProductVariantMutationErrorMock,
  deleteProductVariantMutationMock
} from './variant';
export {
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock
} from './cartItem';
