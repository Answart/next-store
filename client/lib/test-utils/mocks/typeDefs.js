import {
  fakeUser, fakeImage, fakeProduct, fakeVariant, fakeCartItem, fakeOrder, fakeOrderItem,
} from '../utils';


const mockUser = fakeUser();
const mockUser2 = fakeUser();
const mockImage = fakeImage();
const mockProduct = fakeProduct();
const mockVariant = fakeVariant();
const mockCartItem = fakeCartItem();
const mockOrder = fakeOrder();
const mockOrderItem = fakeOrderItem();

mockOrder.items.push(mockOrderItem);

const mockUsers = [
  mockUser,
  {
    ...mockUser2,
    id: '87654321',
    permissions: ['USER'],
  }
];
const mockProducts = [{
  ...mockProduct,
  image: mockImage,
  user: {
    __typename: mockUser.__typename,
    id: mockUser.id,
    name: mockUser.name,
  },
  variants: [{
    ...mockVariant,
    product: {
      __typename: mockProduct.__typename,
      id: mockProduct.id,
      image: mockImage,
    }
  }]
}];

const mockImageVariables = {
  cloudinary_id: mockImage.cloudinary_id,
  name: mockImage.name,
  height: mockImage.height,
  width: mockImage.width,
  transformation: mockImage.transformation,
  image_url: mockImage.image_url,
  large_image_url: mockImage.large_image_url
};

const mockShopProductsVariables = {
  name: mockUser.name,
  orderBy: 'createdAt_DESC',
  skip: 0,
  first: 1
 };


export {
  mockUser, mockUsers,
  mockImage, mockImageVariables,
  mockProduct, mockProducts, mockShopProductsVariables,
  mockVariant,
  mockCartItem,
  mockOrder, mockOrderItem,
};
