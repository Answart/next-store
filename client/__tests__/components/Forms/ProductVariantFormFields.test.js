import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import ProductVariantFormFields from '../../../components/Forms/ProductVariantFormFields';
import { mockImage, mockImageVariables, mockVariant } from '../../../lib/test-utils/mocks';


describe('<ProductVariantFormFields />', () => {
  let wrapper, saveToForm;
  beforeAll(() => {
    wrapper = mount(
      <MockedProvider>
        <ProductVariantFormFields mocks={[]}
          price={0} quantity={0} color='' size='' sale={true} salePrice={0} editView={false} imageIsNew={true}
          image={{
            id: mockImage.id,
            ...mockImageVariables,
            delete_token: 'delete tokeeeen',
          }}
          saveToForm={jest.fn()}
        />
      </MockedProvider>
    );
    saveToForm = wrapper.find('ProductVariantFormFields').props().saveToForm;
  });
  afterAll(() => wrapper.unmount());
  afterEach(() => jest.clearAllMocks());

  it('renders and matches snapshot', async () => {
    const form = wrapper.find('ProductVariantFormFields');
    expect(toJSON(form)).toMatchSnapshot();
  });

  it('renders initial active image', async () => {
    const img = wrapper.find('.form-imgs').find('img');
    expect(img.props().src).toEqual(mockImage.large_image_url);
    expect(img.props().alt).toEqual(mockImage.name);
  });

  describe('input changes', async () => {

    it('size select input works', async () => {
      wrapper.find('#size').simulate('change', { target: {
        value: mockVariant.size, name: 'size', type: 'select',
      }});
      wrapper.update();
      expect(saveToForm).toHaveBeenCalled();
      expect(saveToForm).toHaveBeenCalledWith({ size: mockVariant.size });
    });

    it('color select input works', async () => {
      wrapper.find('#color').simulate('change', { target: {
        value: mockVariant.color, name: 'color', type: 'select',
      }});
      wrapper.update();
      expect(saveToForm).toHaveBeenCalled();
      expect(saveToForm).toHaveBeenCalledWith({ color: mockVariant.color });
    });

    it('quantity input works', async () => {
      wrapper.find('#quantity').simulate('change', { target: {
        value: mockVariant.quantity, name: 'quantity', type: 'number',
      }});
      wrapper.update();
      expect(saveToForm).toHaveBeenCalled();
      expect(saveToForm).toHaveBeenCalledWith({ quantity: mockVariant.quantity });
    });

    it('price input works', async () => {
      wrapper.find('#price').simulate('change', { target: {
        value: mockVariant.price, name: 'price', type: 'number',
      }});
      wrapper.update();
      expect(saveToForm).toHaveBeenCalled();
      expect(saveToForm).toHaveBeenCalledWith({ price: mockVariant.price });
    });

    it('sale input works', async () => {
      wrapper.find('#sale').simulate('change', { target: {
        value: mockVariant.sale, name: 'sale', type: 'checkbox', checked: mockVariant.sale,
      }});
      await wait(0);
      wrapper.update();
      expect(saveToForm).toHaveBeenCalled();
      expect(saveToForm).toHaveBeenCalledWith({ sale: mockVariant.sale, salePrice: 1 });
    });

    it('salePrice input works', async () => {
      const form = wrapper.find('.form-content');
      const instance = wrapper.find('ProductVariantFormFields');
      wrapper.find('#salePrice').simulate('change', { target: {
        value: mockVariant.salePrice, name: 'salePrice', type: 'number',
      }});
      wrapper.update();
      expect(saveToForm).toHaveBeenCalled();
      expect(saveToForm).toHaveBeenCalledWith({ salePrice: mockVariant.salePrice });
    });

    describe('newImage inputs productImage/newImage radio buttons work', async () => {
      it('productImage inputs productImage radio button works', async () => {
        wrapper.find('#productImage').simulate('change', { target: {
          value: 'false', name: 'productImage', type: 'radio', checked: true,
        }});
        await wait(0);
        wrapper.update();
        expect(saveToForm).toHaveBeenCalled();
        expect(saveToForm).toHaveBeenCalledWith({ imageIsNew: false });
      });

      it('newImage inputs productImage radio button works', async () => {
        global.fetch = jest.fn().mockResolvedValue({
          json: () => ({
            success: true,
            message: 'Image deleted.',
          }),
        });
        wrapper.find('#newImage').simulate('change', { target: {
          value: 'true', name: 'newImage', type: 'radio', checked: true,
        }});
        await wait(0);
        wrapper.update();
        expect(saveToForm).toHaveBeenCalled();
        expect(saveToForm).toHaveBeenCalledWith({ imageIsNew: true });
      });
    });

    describe('image/cloudinary upload input', async () => {
      afterEach(() => global.fetch.mockReset())

      it('fails silently if input empty', async () => {
        global.fetch = jest.fn().mockResolvedValue({
          json: () => ({})
        });
        wrapper.find('input[type="file"]').simulate('change', { target: {
          files: [], name: 'upload', type: 'file'
        }});
        await wait(0);
        expect(global.fetch).not.toHaveBeenCalled();
        expect(saveToForm).not.toHaveBeenCalled();
      });

      it('returns alert message when cloudinary upload fails', async () => {
        console.error = jest.fn();
        // mock the global fetch API (invalid res cloudinary)
        global.fetch = jest.fn().mockResolvedValue({
          json: () => ({
            public_id: mockImage.cloudinary_id,
            original_filename: mockImage.name,
            secure_url: mockImage.image_url,
          }),
        });
        wrapper.find('input[type="file"]').simulate('change', { target: {
          files: [mockImage.image_url], name: 'upload', type: 'file',
        }});
        await wait(0);
        expect(global.fetch).toHaveBeenCalled();
        expect(console.error).toHaveBeenCalled();
        expect(console.error).toHaveBeenCalledWith("Error creating image file in cloudinary uploadImageFile. TypeError: Cannot read property 'length' of undefined");
        expect(window.alert).toHaveBeenCalled();
        expect(window.alert).toHaveBeenCalledWith('An error occured while uploading image. Please try again later.');
        expect(saveToForm).not.toHaveBeenCalled();
      });

      describe('successful file upload', async () => {
        beforeEach(() => {
          // mock the global fetch API (cloudinary)
          global.fetch = jest.fn().mockResolvedValue({
            json: () => ({
              public_id: mockImage.cloudinary_id,
              original_filename: mockImage.name,
              secure_url: mockImage.image_url,
              delete_token: 'delete tokeeeen',
              eager: [{
                height: mockImage.height,
                width: mockImage.width,
                transformation: mockImage.transformation,
                secure_url: mockImage.large_image_url,
              }]
            }),
          });
          wrapper.find('input[type="file"]').simulate('change', { target: {
            files: [mockImage.image_url], name: 'upload', type: 'file',
          }});
        });

        it('uploads file to cloudinary and saves returned image to form', async () => {
          await wait(0);
          expect(global.fetch).toHaveBeenCalled();
          expect(saveToForm).toHaveBeenCalled();
        });

        it('calls cloudinary to delete previous image if replacing an already uploaded image', async () => {
          await wait(0);
          wrapper.find('input[type="file"]').simulate('change', { target: {
            files: ['anotherfile.png'], name: 'upload', type: 'file',
          }});
          await wait(0);
          expect(global.fetch).toHaveBeenCalled();
          expect(saveToForm).toHaveBeenCalled();
        });
      });
    })
  });
});
