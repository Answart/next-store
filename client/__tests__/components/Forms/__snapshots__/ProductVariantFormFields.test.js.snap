// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductVariantFormFields /> renders and matches snapshot 1`] = `
<ProductVariantFormFields
  color=""
  editView={false}
  image={
    Object {
      "cloudinary_id": "111111",
      "delete_token": "delete tokeeeen",
      "height": 22,
      "id": "1m4g31d",
      "image_url": "peggswatch1.jpg",
      "large_image_url": "peggswatch2.jpg",
      "name": "peggswatch1",
      "transformation": "",
      "width": 22,
    }
  }
  imageIsNew={true}
  mocks={Array []}
  price={0}
  quantity={0}
  sale={true}
  salePrice={0}
  saveToForm={[MockFunction]}
  size=""
>
  <ProductStyles__StyledProduct>
    <div
      className="ProductStyles__StyledProduct-sc-2231u4-0 dOvnhj"
    >
      <div
        className="form-imgs"
      >
        <img
          alt="peggswatch1"
          height="640"
          src="peggswatch2.jpg"
          width="450"
        />
      </div>
      <div
        className="form-content"
      >
        <div
          className="field-padding"
        >
          <label
            htmlFor="size"
          >
            Size:
            <select
              disabled={false}
              id="size"
              name="size"
              onChange={[Function]}
              value=""
            >
              <option
                key="0"
                value=""
              />
              <option
                key="XXS"
                value="XXS"
              >
                XXS
              </option>
              <option
                key="XS"
                value="XS"
              >
                XS
              </option>
              <option
                key="S"
                value="S"
              >
                S
              </option>
              <option
                key="M"
                value="M"
              >
                M
              </option>
              <option
                key="L"
                value="L"
              >
                L
              </option>
              <option
                key="XL"
                value="XL"
              >
                XL
              </option>
              <option
                key="1X"
                value="1X"
              >
                1X
              </option>
              <option
                key="2X"
                value="2X"
              >
                2X
              </option>
              <option
                key="3X"
                value="3X"
              >
                3X
              </option>
              <option
                key="4X"
                value="4X"
              >
                4X
              </option>
              <option
                key="0"
                value="0"
              >
                0
              </option>
              <option
                key="2"
                value="2"
              >
                2
              </option>
              <option
                key="4"
                value="4"
              >
                4
              </option>
              <option
                key="6"
                value="6"
              >
                6
              </option>
              <option
                key="8"
                value="8"
              >
                8
              </option>
              <option
                key="10"
                value="10"
              >
                10
              </option>
              <option
                key="12"
                value="12"
              >
                12
              </option>
              <option
                key="14"
                value="14"
              >
                14
              </option>
              <option
                key="16"
                value="16"
              >
                16
              </option>
              <option
                key="18"
                value="18"
              >
                18
              </option>
            </select>
          </label>
        </div>
        <div
          className="field-padding"
        >
          <label
            htmlFor="color"
          >
            Color:
            <select
              disabled={false}
              id="color"
              name="color"
              onChange={[Function]}
              value=""
            >
              <option
                key="0"
                value=""
              />
              <option
                key="red"
                value="red"
              >
                red
              </option>
              <option
                key="orange"
                value="orange"
              >
                orange
              </option>
              <option
                key="yellow"
                value="yellow"
              >
                yellow
              </option>
              <option
                key="green"
                value="green"
              >
                green
              </option>
              <option
                key="blue"
                value="blue"
              >
                blue
              </option>
              <option
                key="purple"
                value="purple"
              >
                purple
              </option>
              <option
                key="pink"
                value="pink"
              >
                pink
              </option>
              <option
                key="brown"
                value="brown"
              >
                brown
              </option>
              <option
                key="black"
                value="black"
              >
                black
              </option>
              <option
                key="white"
                value="white"
              >
                white
              </option>
              <option
                key="silver"
                value="silver"
              >
                silver
              </option>
              <option
                key="gold"
                value="gold"
              >
                gold
              </option>
              <option
                key="multi"
                value="multi"
              >
                multi
              </option>
            </select>
          </label>
        </div>
        <div
          className="field-padding"
        >
          <label
            htmlFor="quantity"
          >
            Quantity:
            <input
              id="quantity"
              min="1"
              name="quantity"
              onChange={[Function]}
              placeholder="1"
              required={true}
              type="number"
              value={0}
            />
          </label>
        </div>
        <div
          className="field-padding"
        >
          <label
            htmlFor="price"
          >
            Price per Item (USD):
            <input
              id="price"
              min={1}
              name="price"
              onChange={[Function]}
              placeholder={1}
              required={true}
              step={0.01}
              type="number"
              value={0}
            />
          </label>
        </div>
        <div
          className="field-padding"
        >
          <label
            htmlFor="sale"
          >
            <input
              checked="checked"
              id="sale"
              key="true"
              name="sale"
              onChange={[Function]}
              type="checkbox"
              value={true}
            />
            On Sale
          </label>
        </div>
        <div
          className="field-padding"
        >
          <label
            htmlFor="salePrice"
          >
            Sale Price per Item (USD):
            <input
              disabled={false}
              id="salePrice"
              min={1}
              name="salePrice"
              onChange={[Function]}
              placeholder={1}
              step={0.01}
              type="number"
              value={0}
            />
          </label>
        </div>
        <div
          className="field-padding"
        >
          <label
            htmlFor="image"
          >
            Image:
            <div
              className="prdct-padding"
            >
              <div
                className="prdct-padding"
              >
                <label
                  className="inline-lbl"
                  htmlFor="productImage"
                >
                  <input
                    checked={false}
                    id="productImage"
                    key="false"
                    name="productImage"
                    onChange={[Function]}
                    type="radio"
                    value={false}
                  />
                  Same as Product Image
                </label>
              </div>
              <div>
                <label
                  className="inline-lbl"
                  htmlFor="newImage"
                >
                  <input
                    checked={true}
                    id="newImage"
                    key="true"
                    name="newImage"
                    onChange={[Function]}
                    type="radio"
                    value={true}
                  />
                  New Image
                </label>
              </div>
              <div
                className="field-padding"
              >
                <label
                  className="lbl-button"
                  htmlFor="upload"
                >
                  <input
                    id="upload"
                    name="upload"
                    onChange={[Function]}
                    type="file"
                  />
                  Upload Image
                </label>
              </div>
              <div
                className="field-padding field-detail"
              >
                <h5>
                  Upload Details
                </h5>
                <p>
                  ID: 
                  <i>
                    111111
                  </i>
                </p>
                <p>
                  Name: 
                  <i>
                    peggswatch1
                  </i>
                </p>
                <p>
                  Dimensions: 
                  <i>
                    22
                  </i>
                  W x 
                  <i>
                    22
                  </i>
                  H
                </p>
                <p>
                  Transformation: 
                  <i />
                </p>
                <p>
                  Url: 
                  <i>
                    peggswatch1.jpg
                  </i>
                </p>
              </div>
            </div>
          </label>
        </div>
      </div>
    </div>
  </ProductStyles__StyledProduct>
</ProductVariantFormFields>
`;
